finality-aleph/src/network/substrate.rs

```diff
@@ -60,163 +83,230 @@ impl ProtocolNaming {
+        peer_ids
+            .intersection(&self.connected_peers)
+            .choose(&mut thread_rng())
+            .or_else(|| self.connected_peers.iter().choose(&mut thread_rng()))
+    }
+
+    fn handle_network_event(&mut self, event: SubstrateEvent) -> Option<(Vec<u8>, PeerId)> {
+        use SubstrateEvent::*;
+        match event {
+            ValidateInboundSubstream {
+                peer: _,
+                handshake: _,
+                result_tx,
+            } => {
+                let _ = result_tx.send(ValidationResult::Accept);
+                None
+            }
+            NotificationStreamOpened { peer, .. } => {
+                self.connected_peers.insert(peer);
+                None
+            }
+            NotificationStreamClosed { peer } => {
+                self.connected_peers.remove(&peer);
+                None
+            }
+            NotificationReceived { peer, notification } => Some((notification, peer)),
+        }
+    }
+
+    fn status_report(&self) {
+        let mut status = String::from("Network status report: ");
+        status.push_str(&format!(
+            "{} connected peers - {:?}; ",
+            self.service.protocol(),
+            self.connected_peers.len()
+        ));
+        info!(target: LOG_TARGET, "{}", status);
+    }
+}
+
+#[derive(Debug)]
+pub enum NotificationServiceError {
+    NetworkStreamTerminated,
+}
+
+impl fmt::Display for NotificationServiceError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            NotificationServiceError::NetworkStreamTerminated => {
+                write!(f, "Notifications event stream ended.")
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl<D: Data> GossipNetwork<D> for NotificationService {
+    type Error = NotificationServiceError;
+    type PeerId = PeerId;
+
+    fn send_to(&mut self, data: D, peer_id: PeerId) -> Result<(), Self::Error> {
+        trace!(
+            target: LOG_TARGET,
+            "Sending block sync data to peer {:?}.",
+            peer_id,
+        );
+        self.service.send_sync_notification(&peer_id, data.encode());
+        Ok(())
+    }
+
+    fn send_to_random(&mut self, data: D, peer_ids: HashSet<PeerId>) -> Result<(), Self::Error> {
+        trace!(
+            target: LOG_TARGET,
+            "Sending data to random peer among {:?}.",
+            peer_ids,
+        );
+        let peer_id = match self.random_peer(&peer_ids) {
+            Some(peer_id) => *peer_id,
+            None => {
+                debug!(
+                    target: LOG_TARGET,
+                    "Failed to send message to random peer, no peers are available."
+                );
+                return Ok(());
+            }
+        };
+        self.send_to(data, peer_id)
+    }
+
+    fn broadcast(&mut self, data: D) -> Result<(), Self::Error> {
+        for peer in self.connected_peers.clone() {
+            self.send_to(data.clone(), peer)?;
```

Comment:
I don't think you want to stop at the first fail, no?

Reply:
`send_to` always returns `Ok`, at some point we might want to have a custom broadcast error, but now it's pointless to add the logic supporting it

Reply:
Then an `expect` explaining this would be appropriate here.

Reply:
added `let _ = ...` to ignore all the potential errors for now